### DISCUSSION TITLE
nnx using vmap to create multiple models

### DISCUSSION THREAD

**Discussion by JeyRunner** [google/flax#4048] [Q&A] [answered]
Hi, I want to create multiple instances of a nnx.module (each initialized with a different key).
With linen, I can do:
```python
def make(rng):
  m = my_module.init(rng, dummy_input)
  return ...

rngs = jax.random.split(jax.random.PRNGKey(0), num=5)
models = jax.vmap(make)(rngs)
```

How can I achieve the same with nnx?
I tried:
```python
def make_model(rngs):
  return nnx.Sequential(
    nnx.Linear(..., rngs=rngs),
    ...
  )

init_keys = jax.random.split(jax.random.PRNGKey(0), num=5)
rngs = nnx.Rngs(init_keys)

model = jax.vmap(task.make_model)(rngs)
```
But I get the error `TypeError: object of type 'RngStream' has no len()`.

**Comment by cgarciae** [comment, marked as answer]
EDIT: Updating to use the new APIs.

Hey @JeyRunner! You can use `nnx.split_rngs` to automatically split the `Rngs` before going into `nnx.vmap`.

```python
@nnx.split_rngs(splits=5)
@nnx.vmap
def make_model(rngs):
  return nnx.Linear(2, 3, rngs=rngs)

model = make_model(nnx.Rngs(0))

print(model)
```
Output:
```python
Linear(
  bias=Param(
    value=Array(shape=(5, 3), dtype=float32)
  ),
  bias_init=<function zeros at 0x11ee95f30>,
  dot_general=<function dot_general at 0x11e933910>,
  dtype=None,
  in_features=2,
  kernel=Param(
    value=Array(shape=(5, 2, 3), dtype=float32)
  ),
  kernel_init=<function variance_scaling.<locals>.init at 0x11fa8fe20>,
  out_features=3,
  param_dtype=<class 'jax.numpy.float32'>,
  precision=None,
  use_bias=True
)
```

**Reply by JeyRunner** [reply]
Thanks @cgarciae, thats exactly what I wanted :smiley:  ! 
I didn't know that nxx.vmap can do that.

**Reply by JeyRunner** [reply]
In the meantime I did this:
```python
def make_model(rng):
  rngs = nnx.Rngs(rng)
  return nnx.Linear(2, 3, rngs=rngs)

init_keys = jax.random.split(jax.random.PRNGKey(0), num=5)
model = jax.vmap(task.make_model)(init_keys)
```
This should do kind of the same thing right?

**Reply by cgarciae** [reply]
You must either use [NNX transforms](https://flax.readthedocs.io/en/latest/nnx/nnx_basics.html#nnx-transforms) (nnx.vmap, nnx.jit, nnx.grad, etc) to transform functions that accept Modules or use the [Functional API](https://flax.readthedocs.io/en/latest/nnx/nnx_basics.html#the-functional-api) and handle state manually.

**Comment by errhernandez** [comment]
Hi @cgarciae. In the example that you gave above, namely

def make_model(rngs):
  return nnx.Linear(2, 3, rngs=rngs)

rngs = nnx.Rngs(0)
model = nnx.vmap(make_model, split_rngs=True, axis_size=5)(rngs)

print(model) 

how would you do it if you wanted the dimensions of Linear in make_model to be input?

**Reply by cgarciae** [reply]
Hi @errhernandez, you define new arguments to `make_model` (e.g. `din` and `dout`) and mark them as broadcasts by setting them to `None` in use `in_axes`:
```python
@nnx.split_rngs(splits=5)
@nnx.vmap(in_axes=(None, None, 0))
def make_model(din, dout, rngs):
  return nnx.Linear(din, dout, rngs=rngs)
```

**Reply by errhernandez** [reply]
Thank you @cgarciae for your quick response; your solution works, but now imagine that one wants to have variable dimensions in the stacked nnx.Linear layers; I tried what to me seemed like the obvious extension of your solution, namely
```
@nnx.split_rngs(splits=5)
@nnx.vmap(in_axes(0,0,0))
def make_model(din, dout, rngs):
      return nnx.Linear(din, dout, rngs=rngs)
```
where now din and dout are arrays of length 5. This 'obvious' solution doesn't work: when I try to call make_model I get an error message. 

I am working under the assumption that the aim here is to construct a sequence of nnx.Linear layers, and that an input vector of the right dimensions would go through the first Linear, then the resulting input would go through the second, and so on, right? or am I misunderstanding something?

**Reply by cgarciae** [reply]
Hey @errhernandez, you cannot have Arrays with variable dimensions in JAX. Best you could do is have a list of models, each with different dimension, but you cannot `vmap` over this.

**Reply by errhernandez** [reply]
I feel like I didn't explain myself clearly: it is not that the arrays have variable dimensions; each of the elements in array din would be an input size to the corresponding nnx.Linear, and likewise each of the elements of dout would be the output size; and of course one would need to ensure that dout[i] = din[i+i], but this is the typical situation one has in an MLP, except for the missing activation functions, right? 
So, are you implying that if I construct an MLP like a list of nnx.Linear and activation functions, it is not possible to vmap through it? Can you hint at what would be the way to construct an efficient MLP using nnx?
Sorry for asking so many questions...

**Comment by BeeGass** [comment]
@cgarciae Im trying to understand how to use vmap such that I have multiple models for more than just one dimensions axis. Lets say that we have data that we have restructured to have the shape:
`B, H, T, DH = x.shape`

where:
B: batch
H: heads
T: sequence length
HD: head embed dimension

where we want to parallelize over both the batch dimension (B) and also the head dimension (H). Is this a question of nesting vmaps such that:
```python
@nnx.split_rngs(splits=batch_size)
@nnx.vmap(in_axes=0)
def make_batch_vmap_model(rngs):
    return nnx.Linear(din, dout, rngs=rngs)

@nnx.split_rngs(splits=d_head)
@nnx.vmap(in_axes=0)
def make_d_head_vmap_model(rngs):
    return make_batch_vmap_model(rngs)
```

perhaps this is completely incorrect and that I actually should be using `nnx.pmap`

Thanks in advance for the help

**Reply by BeeGass** [reply]
After looking into more documentation im now realizing that I may not have posed my question correctly, nor with the correct example to highlight it. Im trying to use a combination of pmap and vmap such that I can have multiple models that are vectorized on different hardware. Given this lets consider the scenario where I have `N` devices and `H` heads. Given some input:
`B, T, D = x.shape`

we then rearrange this into:
```python
per_device_batch = B // num_devices
d_head = D // num_heads
x = einops.rearrange(
    x,
    "(nd pdb) t (nh hd) -> nd pdb t nh hd",
    nd=num_devices,
    nbd=per_device_batch,
    nh=num_heads,
    hd=d_head,
) # (num_devices, per_device_batch, seq_len, num_heads, head_dim)
```

Given this, I want to be able to first distribute across my devices, then vectorize over the `per_device_batch` dimension and the `num_heads` dimension on each device for each respective batch. However I have been having trouble accomplishing this. I have tried (and bare with me because I know this might be absurd):
```python
@nnx.split_rngs(splits=num_devices)
@nnx.pmap(in_axes=(0), out_axes=(0))  # device dim
@nnx.split_rngs(splits=per_device_batch)
@nnx.vmap(in_axes=(0), out_axes=(0))  # batch dim
@nnx.split_rngs(splits=num_heads)
@nnx.vmap(in_axes=(1), out_axes=(1))  # head dim
def make_model(rngs):
    return nnx.Linear(din, dout, rngs=rngs)
```

and 

```python
@nnx.split_rngs(splits=num_devices)
@nnx.pmap(in_axes=(0), out_axes=(0))  # device dim
def make_model(rngs):
    @nnx.split_rngs(splits=per_device_batch)
    @nnx.vmap(in_axes=(0), out_axes=(0))  # batch dim
    def make_batched_model(rngs):
        @nnx.split_rngs(splits=num_heads)
        @nnx.vmap(in_axes=(1), out_axes=(1))  # head dim
        def make_n_heads_model(rngs):
            return nnx.Linear(din, dout, rngs=rngs)
        return make_n_heads_model(rngs)
    return make_batched_model(rngs)
```

to no avail. Im likely missing something very obvious which I hope you can help me out with.

**Reply by cgarciae** [reply]
Hey @BeeGass sorry for the late replay.

It works if you specify `in_axes=0` on the last `vmap`:

```python
num_devices = 8
per_device_batch = 4
num_heads = 2
din = 16
dout = 32

@nnx.split_rngs(splits=num_devices)
@nnx.pmap(in_axes=0, out_axes=0)  # device dim
@nnx.split_rngs(splits=per_device_batch)
@nnx.vmap(in_axes=0, out_axes=0)  # batch dim
@nnx.split_rngs(splits=num_heads)
@nnx.vmap(in_axes=0, out_axes=1)  # head dim
def make_model(rngs):
  return nnx.Linear(din, dout, rngs=rngs)

model = make_model(nnx.Rngs(0))
print(model)
```
```
Linear(
  bias=Param(
    value=Array(shape=(8, 4, 32, 2), dtype=float32)
  ),
  bias_init=<function zeros at 0x137c5bac0>,
  dot_general=<function dot_general at 0x127c03a30>,
  dtype=None,
  in_features=16,
  kernel=Param(
    value=Array(shape=(8, 4, 16, 2, 32), dtype=float32)
  ),
  kernel_init=<function variance_scaling.<locals>.init at 0x142594c10>,
  out_features=32,
  param_dtype=<class 'jax.numpy.float32'>,
  precision=None,
  use_bias=True
)
```

**Reply by cgarciae** [reply]
Reason is that the incoming keys on the `Rngs` objects only have one dimension so you cannot vectorize on axis `1`.

**Reply by BeeGass** [reply]
I don't quite understand this. Is this because the axes here are taking in the zeroth `rng`? If this is the case how does one ensure that the model is truly vectorizing over the dimensions described? Is it just then a question of reshaping the input correctly assuming that pmap/vmap takes in the zeroth input, for each?

Also not to worry at all @cgarciae . I appreciate the help.

**Reply by cgarciae** [reply]
No worries, happy to help! The reason is that `split_rngs` takes a key that has shape `()` and turns it into a key with shape `(splits,)`, so it only has available axis `0` for `vmap`.

**Reply by BeeGass** [reply]
> Reason is that the incoming keys on the `Rngs` objects only have one dimension so you cannot vectorize on axis `1`.

Oh I didn't see this. That makes sense. Thank you.

**Reply by BeeGass** [reply]
@cgarciae was working with your suggested answer and realized that initializing the module and it's associated weights in this way is able to create new dimensions, but doesnt seem to vectorize over them when you use the module. I bring this up because I am running into an issue where the new dimensions applied dont interact intuively with the input in the way that one  would as described in your [transformation guide](https://flax.readthedocs.io/en/latest/guides/transforms.html#transforming-methods):

currently im doing a simplified version (that excludes `nnx.pmap`):
```python
batch = 4
num_heads = 2
din = 16
dout = 32

B, T, D = x.shape

@nnx.split_rngs(splits=batch)
@nnx.vmap(in_axes=(None, None, 0), out_axes=0)  
@nnx.split_rngs(splits=num_heads)
@nnx.vmap(in_axes=(None, None, 0), out_axes=0) # <-- moved this to out_axis=0
def make_model(din, dout, rngs):
  return nnx.Linear(din, dout, rngs=rngs)

d_head = D // num_heads
x = einops.rearrange(
    x,
    "b t (nh hd) -> b nh t hd",
    nh=num_heads,
    hd=d_head,
) # (batch, num_heads, seq_len, head_dim)

model = make_model(din, dout, nnx.Rngs(0))

@nnx.vmap(in_axes=0, out_axes=0) # vectorize over inputted batch dim
@nnx.vmap(in_axes=0, out_axes=0) # vectorize over inputted heads dim
def call_model(x):
    return model(x)

x = call_model(x)
```

when I perform this, this errors because:
```
TypeError: dot_general requires contracting dimensions to have the same shape, got (16,) and (4,).
```

which I believe is a result of the kernel dimensions now being `(batch, num_heads, din, dout)` or, given the example we have, `(4, 2, 16, 32)` trying to be multipled with the input that is `(batch, num_heads, seq_len, head_dim)`, or again based on the example, `(4, 2, 128, 16)`. 

This is weird because it was my understanding that while you are using the module that the weights would be in a vectorized state such that when the input actually interacts with the kernel that the dimensions would be `(16, 32)` (instead of what they are currently are, which is `(4, 2, 16, 32)`). Clearly this isnt the case but I wonder how to enable this. It seems like its possible based on the guide but im not clear how this would be performed. Im hoping I could ask for your help again.

**Reply by BeeGass** [reply]
I was able to figure it out. The issue is that when you go to call the model you need to vectorize over the model and the parameter dimensions itself. More concretely, here is an example of the same problem arising because I was not vectorizing over the model's parameter dimensions, of choice:
```python
import jax
import jax.numpy as jnp
from flax import nnx
import einops

# Configuration
batch = 4
num_heads = 2
din = 16  # input dimension
dout = 32 # output dimension
seq_len = 8

# Create sample input
x = jnp.ones((batch, seq_len, din))  # (batch, seq_len, din)

# Reshape input to (batch, heads, seq_len, head_dim)
d_head = din // num_heads
x = einops.rearrange(
    x,
    "b t (nh hd) -> b nh t hd",
    nh=num_heads,
    hd=d_head,
)
print("Input shape:", x.shape)

# Model initialization attempt with nested vmaps
@nnx.split_rngs(splits=batch)
@nnx.vmap(in_axes=(None, None, 0), out_axes=0)  # batch dim
@nnx.split_rngs(splits=num_heads)
@nnx.vmap(in_axes=(None, None, 0), out_axes=0)  # head dim
def make_model(din, dout, rngs):
    return nnx.Linear(din, dout, rngs=rngs)

# Create model
model = make_model(din, dout, nnx.Rngs(0))

# Print model parameters shape
print("\nModel parameter shapes:")
print("Kernel:", model.kernel.value.shape)
print("Bias:", model.bias.value.shape)

# Attempt to use the model
@nnx.vmap(in_axes=0, out_axes=0)  # batch dim
@nnx.vmap(in_axes=0, out_axes=0)  # head dim
def call_model(x):
    return model(x)

try:
    output = call_model(x)
    print("\nOutput shape:", output.shape)
except Exception as e:
    print("\nError when calling model:")
    print(e)
```

This will produce a similar error as above where:
```
dot_general requires contracting dimensions to have the same shape, got (8,) and (4,).
```

however if we ensure that we can vmap over the model's parameters we can avoid this behavior:
```python
import jax
import jax.numpy as jnp
from flax import nnx

# Configuration
batch = 2
num_heads = 3
head_dim = 4  # input dimension per head
dout = 5      # output dimension
seq_len = 6

# Initialize vectorized model
@nnx.split_rngs(splits=batch)
@nnx.vmap(in_axes=0)  # batch dimension
@nnx.split_rngs(splits=num_heads) 
@nnx.vmap(in_axes=0)  # head dimension
def make_model(rngs):
    # Each Linear layer processes head_dim -> dout
    return nnx.Linear(head_dim, dout, rngs=rngs)

model = make_model(nnx.Rngs(0))

# Create sample input (batch, num_heads, seq_len, head_dim)
x = jnp.ones((batch, num_heads, seq_len, head_dim))

# Forward pass with vmap
def forward(model, x):
    @nnx.vmap(in_axes=(0, 0))  # batch dimension
    @nnx.vmap(in_axes=(0, 0))  # head dimension 
    def forward_single(model, x):
        # x shape: (seq_len, head_dim)
        # model.kernel shape: (head_dim, dout)
        return model(x)

    return forward_single(model, x)

# Run and print shapes
print("Model parameter shapes:")
print(f"Kernel shape: {model.kernel.value.shape}")  # Should be (batch, num_heads, head_dim, dout)
print(f"Bias shape: {model.bias.value.shape}\n")   # Should be (batch, num_heads, dout)

print("Input/Output shapes:")
print(f"Input shape: {x.shape}")
output = forward(model, x)
print(f"Output shape: {output.shape}")
```

however this becomes a little less intuitive when you go about having a module/block that calls many different modules but you still want to vectorize over this. I have dont have a good toy example for this unfortunately. I was playing around with a block that was a linear layer feeding into an RNN where I wanted batch number and head number versions of this block. My hope was that I could vectorize over this block and you can! This is done in a simialar behavior I describe above. During the process of initializing the block where you do the initializing vmaps this creates extra dimensions of the parameters within the specfic module within the block which you then implicitly vmap over by vmapping over the block module when you do the "call" vmaps. 

Hopefully this makes sense. If it doesnt, please ping me and I can try to think a bit harder about making a toy example.
